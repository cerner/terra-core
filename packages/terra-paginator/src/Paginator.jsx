import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ResponsiveElement from 'terra-responsive-element';
import KeyCode from 'keycode-js';
import styles from './Paginator.module.scss';
import { calculatePages, pageSet } from './_paginationUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Function to be executed when a navigation element is selected.
   */
  onPageChange: PropTypes.func.isRequired,
  /**
   * The active/selected page. Used to set the default selected page or maintain selection across renders.
   * Required when using totalCount and itemCountPerPage.
   */
  selectedPage: PropTypes.number,
  /**
   * Total number of all items being paginated.
   * Required when using itemCountPerPage and selectedPage.
   */
  totalCount: PropTypes.number,
  /**
   * Total number of items per page.
   * Required when using selectedPage and totalCount.
   */
  itemCountPerPage: PropTypes.number,
};

class Paginator extends React.Component {
  constructor(props) {
    super(props);

    const { selectedPage, totalCount, itemCountPerPage } = this.props;

    this.state = {
      selectedPage: selectedPage && totalCount ? selectedPage : undefined,
      pageSequence: selectedPage && totalCount ? pageSet(selectedPage, calculatePages(totalCount, itemCountPerPage)) : undefined,
    };

    this.handlePageChange = this.handlePageChange.bind(this);
    this.handleOnKeyDown = this.handleOnKeyDown.bind(this);
    this.hasNavContext = this.hasNavContext.bind(this);
    this.buildPageButtons = this.buildPageButtons.bind(this);
    this.reducedPaginator = this.reducedPaginator.bind(this);
  }

  handlePageChange(index) {
    return (event) => {
      event.preventDefault();

      if (Number.isNaN(Number(index))) {
        this.props.onPageChange(event.currentTarget.attributes['aria-label'].value);

        return false;
      }

      this.props.onPageChange(index);
      this.setState({
        selectedPage: index,
        pageSequence: pageSet(index, calculatePages(this.props.totalCount, this.props.itemCountPerPage)),
      });

      return false;
    };
  }

  handleOnKeyDown(index) {
    return (event) => {
      if (event.nativeEvent.keyCode === KeyCode.KEY_RETURN || event.nativeEvent.keyCode === KeyCode.KEY_SPACE) {
        event.preventDefault();

        if (Number.isNaN(Number(index))) {
          this.props.onPageChange(event.currentTarget.attributes['aria-label'].value);

          return false;
        }

        this.props.onPageChange(index);
        this.setState({
          selectedPage: index,
          pageSequence: pageSet(index, calculatePages(this.props.totalCount, this.props.itemCountPerPage)),
        });
      }

      return false;
    };
  }

  // TODO: Resolve lint issues - https://github.com/cerner/terra-core/issues/1689
  /* eslint-disable jsx-a11y/no-static-element-interactions, jsx-a11y/anchor-is-valid, jsx-a11y/no-noninteractive-tabindex */
  buildPageButtons(totalPages, onClick) {
    const { pageSequence, selectedPage } = this.state;

    const pageButtons = [];

    pageSequence.forEach((val) => {
      const paginationLinkClassNames = cx([
        'nav-link',
        val === selectedPage ? 'is-selected' : null,
      ]);

      if (val > totalPages) {
        return;
      }
      /* eslint-disable jsx-a11y/aria-props */
      pageButtons.push((
        <a
          aria-label={`Page ${val}`}
          aria-current={val === selectedPage && 'page'}
          className={paginationLinkClassNames}
          tabIndex={val === selectedPage ? null : '0'}
          key={`pageButton_${val}`}
          onClick={onClick(val)}
          onKeyDown={this.handleOnKeyDown(val)}
        >
          {val}
        </a>
      ));
      /* eslint-enable jsx-a11y/aria-props */
    });

    return pageButtons;
  }

  hasNavContext() {
    return this.props.totalCount && this.props.itemCountPerPage;
  }

  defaultPaginator() {
    const totalPages = calculatePages(this.props.totalCount, this.props.itemCountPerPage);
    const { selectedPage } = this.state;
    const previousPageIndex = selectedPage === 1 ? 1 : selectedPage - 1;
    const nextPageIndex = selectedPage === totalPages ? totalPages : selectedPage + 1;

    const fullView = (
      <div className={cx(['paginator', !this.hasNavContext() && 'pageless'])}>
        {
          this.hasNavContext() && (
          <a
            aria-disabled={selectedPage === 1}
            aria-label="first"
            className={cx(['nav-link', 'left-controls', selectedPage === 1 && 'is-disabled'])}
            tabIndex={selectedPage === 1 ? null : '0'}
            onClick={this.handlePageChange(1)}
            onKeyDown={this.handleOnKeyDown(1)}
          >
            First
          </a>
          )
        }
        <a
          aria-disabled={selectedPage === 1}
          aria-label="previous"
          className={cx(['nav-link', 'left-controls', 'previous', selectedPage === 1 && 'is-disabled'])}
          tabIndex={selectedPage === 1 ? null : '0'}
          onClick={this.handlePageChange(previousPageIndex)}
          onKeyDown={this.handleOnKeyDown(previousPageIndex)}
        >
          <span className={cx('icon')} />
Previous
        </a>
        {this.hasNavContext() && this.buildPageButtons(totalPages, this.handlePageChange)}
        <a
          aria-disabled={selectedPage === totalPages}
          aria-label="next"
          className={cx(['nav-link', 'right-controls', 'next', selectedPage === totalPages && 'is-disabled'])}
          tabIndex={selectedPage === totalPages ? null : '0'}
          onClick={this.handlePageChange(nextPageIndex)}
          onKeyDown={this.handleOnKeyDown(nextPageIndex)}
        >
          Next
          <span className={cx('icon')} />
        </a>
        {
          this.hasNavContext() && (
          <a
            aria-disabled={selectedPage === totalPages}
            aria-label="last"
            className={cx(['nav-link', 'right-controls', selectedPage === totalPages && 'is-disabled'])}
            tabIndex={selectedPage === totalPages ? null : '0'}
            onClick={this.handlePageChange(totalPages)}
            onKeyDown={this.handleOnKeyDown(totalPages)}
          >
            Last
          </a>
          )
        }
      </div>
    );

    return fullView;
  }

  reducedPaginator() {
    const totalPages = calculatePages(this.props.totalCount, this.props.itemCountPerPage);
    const { selectedPage } = this.state;
    const previousPageIndex = selectedPage === 1 ? 1 : selectedPage - 1;
    const nextPageIndex = selectedPage === totalPages ? totalPages : selectedPage + 1;

    const reducedView = (
      <div className={cx(['paginator', !this.hasNavContext() && 'pageless'])} role="navigation" aria-label="pagination">
        {
          this.hasNavContext() && (
          <a
            aria-disabled={selectedPage === 1}
            aria-label="first"
            className={cx(['nav-link', 'left-controls', selectedPage === 1 && 'is-disabled'])}
            tabIndex={selectedPage === 1 ? null : '0'}
            onClick={this.handlePageChange(1)}
            onKeyDown={this.handleOnKeyDown(1)}
          >
            First
          </a>
          )
        }
        <a
          aria-disabled={selectedPage === 1}
          aria-label="previous"
          className={cx(['nav-link', 'left-controls', 'previous', 'icon-only', selectedPage === 1 && 'is-disabled'])}
          tabIndex={selectedPage === 1 ? null : '0'}
          onClick={this.handlePageChange(previousPageIndex)}
          onKeyDown={this.handleOnKeyDown(previousPageIndex)}
        >
          <span className={cx('visually-hidden')}>Previous</span>
          <span className={cx('icon')} />
        </a>
        {this.hasNavContext() && `Page ${selectedPage}`}
        <a
          aria-disabled={selectedPage === totalPages}
          aria-label="next"
          className={cx(['nav-link', 'right-controls', 'next', 'icon-only', selectedPage === totalPages && 'is-disabled'])}
          tabIndex={selectedPage === totalPages ? null : '0'}
          onClick={this.handlePageChange(nextPageIndex)}
          onKeyDown={this.handleOnKeyDown(nextPageIndex)}
        >
          <span className={cx('visually-hidden')}>Next</span>
          <span className={cx('icon')} />
        </a>
        {
          this.hasNavContext() && (
          <a
            aria-disabled={selectedPage === totalPages}
            aria-label="last"
            className={cx(['nav-link', 'right-controls', selectedPage === totalPages && 'is-disabled'])}
            tabIndex={selectedPage === totalPages ? null : '0'}
            onClick={this.handlePageChange(totalPages)}
            onKeyDown={this.handleOnKeyDown(totalPages)}
          >
            Last
          </a>
          )
        }
      </div>
    );

    return reducedView;
  }

  /* eslint-enable jsx-a11y/no-static-element-interactions, jsx-a11y/anchor-is-valid, jsx-a11y/no-noninteractive-tabindex */
  render() {
    return <ResponsiveElement tiny={this.reducedPaginator()} medium={this.defaultPaginator()} />;
  }
}

Paginator.propTypes = propTypes;

export default Paginator;
