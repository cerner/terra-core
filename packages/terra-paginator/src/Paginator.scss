@import '~terra-mixins';

:local {
  .paginator {
    align-items: center;
    display: flex;
    justify-content: center;

    &.pageless {
      justify-content: space-between;
    }

    &.progressive {
      justify-content: space-between;
    }

    .nav-link {
      background: var(--terra-paginator-nav-link-background);
      border: var(--terra-paginator-nav-link-border);
      border-radius: var(--terra-paginator-nav-link-border-radius);
      color: var(--terra-paginator-nav-link-foreground-color, #0059a8);
      cursor: pointer;
      display: inline-block;
      margin-left: var(--terra-paginator-nav-link-margin-left, 2px);
      margin-right: var(--terra-paginator-nav-link-margin-right, 2px);
      min-width: var(--terra-paginator-nav-link-min-width, 2rem); // Helps ensure limited resizing as you page through large numbers.
      padding: var(--terra-paginator-nav-link-padding, 3px);
      text-align: center;
      user-select: none; // Keeps user from highlighting the paging link text.

      &:hover {
        color: var(--terra-paginator-nav-link-foreground-color-hover, #001f67);
        text-decoration: underline;
      }

      &:focus {
        background: var(--terra-paginator-nav-link-background-focus);
      }

      &:active {
        color: var(--terra-paginator-nav-link-foreground-color-active, #0059a8);
        text-decoration: var(--terra-paginator-nav-link-text-decoration-active);
      }
    }

    .is-selected {
      color: initial;
      cursor: initial;
      pointer-events: none;

      &:hover {
        text-decoration: none;
      }

      &:focus {
        outline: none;
      }
    }

    .is-disabled {
      color: var(--terra-paginator-nav-link-foreground-color-disabled, rgba(28, 31, 33, 0.2));
      cursor: default;
      pointer-events: none;
    }

    .left-controls {
      margin-left: var(--terra-paginator-nav-link-margin-left-left-controls, 2px);
      margin-right: var(--terra-paginator-nav-link-margin-right-left-controls, 8px);
    }

    .right-controls {
      margin-left: var(--terra-paginator-nav-link-margin-left-left-controls, 8px);
      margin-right: var(--terra-paginator-nav-link-margin-right-left-controls, 2px);
    }

    .icon {
      display: inline-block; // IE 10
      height: 1rem;
      vertical-align: middle;
      width: 1rem;
    }

    &.previous > .icon {
      content: var(--terra-paginator-nav-link-icon-previous, inline-svg('<svg viewBox="0 0 48 48" xmlns="http://www.w3.org/2000/svg"><path fill="#0059a8" d="M10.3 24L33.8 0l3.9 3.8L18 24l19.7 20.2-3.9 3.8z"></path></svg>'));
    }

    &.previous.is-disabled > .icon {
      content: var(--terra-paginator-nav-link-icon-previous-disabled, inline-svg('<svg viewBox="0 0 48 48" xmlns="http://www.w3.org/2000/svg"><path fill="rgba(28, 31, 33, 0.2)" d="M10.3 24L33.8 0l3.9 3.8L18 24l19.7 20.2-3.9 3.8z"></path></svg>'));
    }

    &.next > .icon {
      content: var(--terra-paginator-nav-link-icon-next, inline-svg('<svg viewBox="0 0 48 48" xmlns="http://www.w3.org/2000/svg"><path fill="#0059a8" d="M37.7 24L14.2 48l-3.9-3.8L30 24 10.3 3.8 14.2 0z"></path></svg>'));
    }

    &.next.is-disabled > .icon {
      content: var(--terra-paginator-nav-link-icon-next-disabled, inline-svg('<svg viewBox="0 0 48 48" xmlns="http://www.w3.org/2000/svg"><path fill="rgba(28, 31, 33, 0.2)" d="M37.7 24L14.2 48l-3.9-3.8L30 24 10.3 3.8 14.2 0z"></path></svg>'));
    }
  }
}
