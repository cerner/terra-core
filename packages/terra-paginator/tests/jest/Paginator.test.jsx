import React from 'react';
import ThemeContextProvider from 'terra-theme-context/lib/ThemeContextProvider';

/* eslint-disable-next-line import/no-extraneous-dependencies */
import { shallowWithIntl, mountWithIntl } from 'terra-enzyme-intl';
import Paginator from '../../src/Paginator';

describe('Paginator', () => {
  const defaultRender = <Paginator onPageChange={e => typeof e} selectedPage={1} totalCount={2234} itemCountPerPage={20} />;
  const noPagesRender = <Paginator onPageChange={e => typeof e} />;

  // Snapshot Tests
  it('should render a default paginator - Tiny', () => {
    const wrapper = shallowWithIntl(defaultRender).dive();
    wrapper.setState({ breakpoint: 'tiny' });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  // Snapshot Tests
  it('should render a default paginator - Large', () => {
    const wrapper = shallowWithIntl(defaultRender).dive();
    wrapper.setState({ breakpoint: 'large' });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  // Snapshot Tests
  it('should render a paginator without page links - Tiny', () => {
    const wrapper = shallowWithIntl(noPagesRender).dive();
    wrapper.setState({ breakpoint: 'tiny' });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  // Snapshot Tests
  it('should render a paginator without page links - Large', () => {
    const wrapper = shallowWithIntl(noPagesRender).dive();
    wrapper.setState({ breakpoint: 'large' });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  // Snapshot Tests
  it('correctly applies the theme context className', () => {
    const wrapper = mountWithIntl(
      <ThemeContextProvider theme={{ className: 'orion-fusion-theme' }}>
        {noPagesRender}
      </ThemeContextProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
