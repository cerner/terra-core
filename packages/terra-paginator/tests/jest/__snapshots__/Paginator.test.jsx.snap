// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator should render a default paginator 1`] = `
<ResponsiveElement
  medium={
    <div
      className="paginator"
    >
      <a
        aria-disabled={true}
        aria-label="first"
        className="nav-link left-controls is-disabled"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={null}
      >
        First
      </a>
      <a
        aria-disabled={true}
        aria-label="previous"
        className="nav-link left-controls previous is-disabled"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={null}
      >
        <span
          className="icon"
        />
        Previous
      </a>
      <a
        aria-current="page"
        aria-label="Page 1"
        className="nav-link is-selected"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={null}
      >
        1
      </a>
      <a
        aria-current={false}
        aria-label="Page 2"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        2
      </a>
      <a
        aria-current={false}
        aria-label="Page 3"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        3
      </a>
      <a
        aria-current={false}
        aria-label="Page 4"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        4
      </a>
      <a
        aria-current={false}
        aria-label="Page 5"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        5
      </a>
      <a
        aria-current={false}
        aria-label="Page 6"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        6
      </a>
      <a
        aria-current={false}
        aria-label="Page 7"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        7
      </a>
      <a
        aria-current={false}
        aria-label="Page 8"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        8
      </a>
      <a
        aria-current={false}
        aria-label="Page 9"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        9
      </a>
      <a
        aria-current={false}
        aria-label="Page 10"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        10
      </a>
      <a
        aria-disabled={false}
        aria-label="next"
        className="nav-link right-controls next"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        Next
        <span
          className="icon"
        />
      </a>
      <a
        aria-disabled={false}
        aria-label="last"
        className="nav-link right-controls"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        Last
      </a>
    </div>
  }
  responsiveTo="parent"
  tiny={
    <div
      aria-label="pagination"
      className="paginator"
      role="navigation"
    >
      <a
        aria-disabled={true}
        aria-label="first"
        className="nav-link left-controls is-disabled"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={null}
      >
        First
      </a>
      <a
        aria-disabled={true}
        aria-label="previous"
        className="nav-link left-controls previous icon-only is-disabled"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={null}
      >
        <span
          className="visually-hidden"
        >
          Previous
        </span>
        <span
          className="icon"
        />
      </a>
      Page 1
      <a
        aria-disabled={false}
        aria-label="next"
        className="nav-link right-controls next icon-only"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        <span
          className="visually-hidden"
        >
          Next
        </span>
        <span
          className="icon"
        />
      </a>
      <a
        aria-disabled={false}
        aria-label="last"
        className="nav-link right-controls"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        Last
      </a>
    </div>
  }
/>
`;

exports[`Paginator should render a paginator without page links 1`] = `
<ResponsiveElement
  medium={
    <div
      className="paginator pageless"
    >
      <a
        aria-disabled={false}
        aria-label="previous"
        className="nav-link left-controls previous"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        <span
          className="icon"
        />
        Previous
      </a>
      <a
        aria-disabled={false}
        aria-label="next"
        className="nav-link right-controls next"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        Next
        <span
          className="icon"
        />
      </a>
    </div>
  }
  responsiveTo="parent"
  tiny={
    <div
      aria-label="pagination"
      className="paginator pageless"
      role="navigation"
    >
      <a
        aria-disabled={false}
        aria-label="previous"
        className="nav-link left-controls previous icon-only"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        <span
          className="visually-hidden"
        >
          Previous
        </span>
        <span
          className="icon"
        />
      </a>
      <a
        aria-disabled={false}
        aria-label="next"
        className="nav-link right-controls next icon-only"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
      >
        <span
          className="visually-hidden"
        >
          Next
        </span>
        <span
          className="icon"
        />
      </a>
    </div>
  }
/>
`;
